airflow/providers/amazon/aws/hooks/batch_client.py

```diff
@@ -0,0 +1,503 @@
+        10 sec +/- 5 sec: ``delay = AwsBatchClient.add_jitter(10, width=5, minima=0)``
+
+    .. seealso::
+        - https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/batch.html
+        - https://docs.aws.amazon.com/general/latest/gr/api-retries.html
+        - https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+    """
+
+    MAX_RETRIES = 4200
+    STATUS_RETRIES = 10
+
+    # delays are in seconds
+    DEFAULT_DELAY_MIN = 1
+    DEFAULT_DELAY_MAX = 10
+
+    def __init__(
+        self, max_retries=None, status_retries=None, aws_conn_id=None, region_name=None
+    ):
+        super().__init__()
+        self.max_retries = max_retries or self.MAX_RETRIES
+        self.status_retries = status_retries or self.STATUS_RETRIES
+        self.aws_conn_id = aws_conn_id
+        self.region_name = region_name
+        self._hook = None
+        self._client = None
+
+    @property
+    def hook(self):
+        """An AWS API connection manager (wraps boto3)
+
+        :rtype: AwsHook
+        """
+        if self._hook is None:
+            self._hook = AwsHook(aws_conn_id=self.aws_conn_id)
+        return self._hook
+
+    @property
+    def client(self):
+        """An AWS API client for batch services, like ``boto3.client('batch')``
+
+        :rtype: Union[BatchProtocol, botocore.client.BaseClient]
+        """
+        if self._client is None:
+            self._client = self.hook.get_client_type("batch", region_name=self.region_name)
+        return self._client
+
+    def terminate_job(self, job_id, reason):
+        """Terminate a batch job
+
+        :param job_id: a jobId to terminate
+        :type job_id: str
+
+        :param reason: a reason to terminate jobId
+        :type reason: str
+
+        :return: an API response
+        :rtype: dict
+        """
+        response = self.client.terminate_job(jobId=job_id, reason=reason)
+        self.log.info(response)
+        return response
+
+    def check_job_success(self, job_id):
+        """
+        Check the final status of the batch job; return True if the job
+        'SUCCEEDED', else raise an AirflowException
+
+        :param job_id: a batch job ID
+        :type job_id: str
+
+        :raises: AirflowException
+        :rtype: bool
+        """
+        job = self.get_job_description(job_id)
+        job_status = job.get("status")
+        if job_status == "SUCCEEDED":
+            self.log.info("AWS batch job ({}) succeeded: {}".format(job_id, job))
+            return True
+        elif job_status == "FAILED":
+            reason = "AWS batch job ({}) failed: {}".format(job_id, job)
+        elif job_status in ["SUBMITTED", "PENDING", "RUNNABLE", "STARTING", "RUNNING"]:
+            reason = "AWS batch job ({}) is not complete: {}".format(job_id, job)
+        else:
+            reason = "AWS batch job ({}) has unknown status: {}".format(job_id, job)
+        raise AirflowException(reason)
```

Comment:
```python
        if job_status == "SUCCEEDED":
            self.log.info("AWS batch job ({}) succeeded: {}".format(job_id, job))
            return True

        if job_status == "FAILED":
            raise AirflowException("AWS batch job ({}) failed: {}".format(job_id, job))

        if job_status in ["SUBMITTED", "PENDING", "RUNNABLE", "STARTING", "RUNNING"]:
            raise AirflowException("AWS batch job ({}) is not complete: {}".format(job_id, job))

        raise AirflowException("AWS batch job ({}) has unknown status: {}".format(job_id, job))
```

That's how I would do it. Yours is fine, too :)

Reply:
I like the `if return/raise`, `if return/raise` more (personally).

Reply:
Thanks, I also prefer the extra space and separation of control-flow in the suggested changes.  It reads better as isolated control-flow, rather than the appearance of integrated flow.  The point of integration is the job-status string, so it makes sense from that perspective to have an `if ... elsif ... else` flow, but the use of return/raise is clearer in the suggested pattern.  I have no idea whether there's any performance differences and don't really care for something that's only called once at the end of a batch job.  Anyhow, suggested change adopted.